dali/operators/sequence/sequence_rearrange.cc

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2019-2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vector>
+
+#include "dali/operators/sequence/sequence_rearrange.h"
+
+namespace dali {
+
+
+DALI_SCHEMA(SequenceRearrange)
+    .DocStr(R"code(Rearrange the sequence stored as tensor.
+Assumes that the outermost dimension represents a sequence and other dimensions of input
+represent elements of that sequence. If layout is specified, the first dimension should
+be denoted as ``F`` indicating frames of the sequence.)code")
+    .NumInput(1)
+    .NumOutput(1)
+    .AllowSequences()
+    .AddArg("new_order", R"code(List describing new order for elements of each sample.
+Output sequence at position ``i`` will contain element ``new_order[i]`` from input sequence.
+Elements can be repeated or dropped, empty output sequences are not allowed.
+Only indices in ``[0, input_outermost_dimension)`` are allowed
+to be used in ``new_order``. Can be specified per sample as 1D tensors.)code",
+            DALI_INT_VEC, true);
+
+TensorShape<> GetOutputShape(const TensorShape<> &in_sample_shape,
+                             const TensorView<StorageCPU, const int, 1> &new_order,
+                             int sample_idx) {
+  const int in_seq_length = GetSeqLength(in_sample_shape);
+  const int out_seq_length = new_order.num_elements();
+  for (int i = 0; i < out_seq_length; i++) {
+    int src_idx = new_order.data[i];
+    DALI_ENFORCE(
+        0 <= src_idx && src_idx < in_seq_length,
+        make_string("Source element src_idx must be between 0 and input_sequence_length = ",
+                    in_seq_length, " for sample ", sample_idx, ", but it is: ", src_idx, "."));
+  }
+  auto element_shape = in_sample_shape.last(in_sample_shape.sample_dim() - 1);
+  auto new_sample_shape = shape_cat(out_seq_length, element_shape);
```

Comment:
I think this one is easier to understand:
```suggestion
  auto new_sample_shape = in_sample_shape;
  new_sample_shape[0] = out_seq_length;
```